### ISSUE TITLE
TypeError on from_dict on dataclass with nested dict and tuple

### ISSUE THREAD

**Issue by christian-steinmeyer** [lidatong/dataclasses-json#294]
When creating a dataclass with a nested dict and a tuple, `from_dict()` does not work.

For a minimal working example:
```python
from typing import Dict, Tuple
from dataclasses import dataclass
from dataclasses_json import DataClassJsonMixin

@dataclass
class TestClass(DataClassJsonMixin):
    nested_dict: Dict[str, Dict[Tuple[str], int]]

my_test = TestClass.from_dict({"nested_dict": {"a":{("b", "c"): 1, "c": 2}, "d": {}}})
```

This yields:
```bash
Traceback (most recent call last):
  File "[PROJECT_PATH]python3.8/site-packages/dataclasses_json/core.py", line 263, in _decode_generic
    res = _get_type_cons(type_)(xs)
  File "/usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/typing.py", line 727, in __call__
    raise TypeError(f"Type {self._name} cannot be instantiated; "
TypeError: Type Tuple cannot be instantiated; use tuple() instead
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "[PROJECT_PATH]python3.8/site-packages/dataclasses_json/core.py", line 263, in _decode_generic
    res = _get_type_cons(type_)(xs)
  File "[PROJECT_PATH]python3.8/site-packages/dataclasses_json/core.py", line 309, in <genexpr>
    items = (_decode_generic(type_arg, x, infer_missing) for x in xs)
  File "[PROJECT_PATH]python3.8/site-packages/dataclasses_json/core.py", line 265, in _decode_generic
    res = type_(xs)
  File "/usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/typing.py", line 727, in __call__
    raise TypeError(f"Type {self._name} cannot be instantiated; "
TypeError: Type Dict cannot be instantiated; use dict() instead
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "[PROJECT_PATH]python3.8/site-packages/IPython/core/interactiveshell.py", line 3437, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-15-a734de727315>", line 1, in <module>
    bob = TestClass.from_dict({"nested_dict": {"a":{("b", "c"): 1, "c": 2}, "d": {}}})
  File "[PROJECT_PATH]python3.8/site-packages/dataclasses_json/api.py", line 83, in from_dict
    return _decode_dataclass(cls, kvs, infer_missing)
  File "[PROJECT_PATH]python3.8/site-packages/dataclasses_json/core.py", line 201, in _decode_dataclass
    init_kwargs[field.name] = _decode_generic(field_type,
  File "[PROJECT_PATH]python3.8/site-packages/dataclasses_json/core.py", line 265, in _decode_generic
    res = type_(xs)
  File "/usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/typing.py", line 727, in __call__
    raise TypeError(f"Type {self._name} cannot be instantiated; "
TypeError: Type Dict cannot be instantiated; use dict() instead
``` 

Is there anything I am obviously doing wrong or a workaround without getting rid of the type annotations?

**Comment by lidatong** [comment]
Hi, thanks for reporting this and providing the example & error.

I'll have to look more closely to see why this is happening -- there is code to handle `Dict` types as fields, but perhaps it's something to do with nested dicts

> Is there anything I am obviously doing wrong or a workaround without getting rid of the type annotations?

A workaround for now would be to construct the dataclass manually. Sorry for the inconvenience. I'll update if with more details / a fix

**Comment by christian-steinmeyer** [comment]
Thanks for the quick reply!

When I try to manually construct to create such a class, I have another problem in the other direction as well:
continuing from the example above and assuming that I constructed `my_test` some other way, when I now call
```
my_test.to_dict()
```

I get the error that `List is not hashable`. After debugging for a bit, I found the root cause for that to be [this cast to list](https://github.com/lidatong/dataclasses-json/blob/ccff0b51418b4c41b63ee48f31126435e5cfd2bd/dataclasses_json/core.py#L336). The tuple that functions as my dictionary key is being picked up (correctly) as a collection, but due to the cast the surrounding dictionary now has trouble adding the resulting list as a key. This, unfortunately for me, all makes sense. I'm afraid the most reasonable thing to do is a wrapper class for my collection that implements a hash method or do you have a better idea?

**Comment by pearceeverydaylabs** [comment]
I attempted a fix. Feedback welcome :) https://github.com/lidatong/dataclasses-json/pull/297

**Comment by pearceeverydaylabs** [comment]
@christian-steinmeyer can this be closed now that #297 is merged?

**Comment by christian-steinmeyer** [comment]
Yes, thanks!
